nexus/src/app/sagas/snapshot_create.rs

```diff
@@ -0,0 +1,765 @@
+        .await?;
+
+    Ok(())
+}
+
+async fn ssc_send_snapshot_request(
+    sagactx: ActionContext<SagaSnapshotCreate>,
+) -> Result<(), ActionError> {
+    let log = sagactx.user_data().log();
+    let osagactx = sagactx.user_data();
+    let params = sagactx.saga_params();
+    let opctx = OpContext::for_saga_action(&sagactx, &params.serialized_authn);
+
+    let snapshot_id = sagactx.lookup::<Uuid>("snapshot_id")?;
+
+    // Find if this disk is attached to an instance
+    let (.., disk) = LookupPath::new(&opctx, &osagactx.datastore())
+        .project_id(params.project_id)
+        .disk_name(&params.create_params.disk.clone().into())
+        .fetch()
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    match disk.runtime().attach_instance_id {
+        Some(instance_id) => {
+            warn!(log, "disk {} instance is {}", disk.id(), instance_id);
+
+            // Get the instance's sled agent client
+            let (.., instance) = LookupPath::new(&opctx, &osagactx.datastore())
+                .instance_id(instance_id)
+                .fetch()
+                .await
+                .map_err(ActionError::action_failed)?;
+
+            let sled_agent_client = osagactx
+                .nexus()
+                .instance_sled(&instance)
+                .await
+                .map_err(ActionError::action_failed)?;
+
+            warn!(log, "instance {} sled agent created ok", instance_id);
+
+            // Send a snapshot request to propolis through sled agent
+            sled_agent_client
+                .instance_issue_disk_snapshot_request(
+                    &instance.id(),
+                    &disk.id(),
+                    &InstanceIssueDiskSnapshotRequestBody { snapshot_id },
+                )
+                .await
+                .map_err(|e| e.to_string())
+                .map_err(ActionError::action_failed)?;
+        }
+
+        None => {
+            warn!(log, "disk {} not attached to an instance", disk.id());
+
+            // Grab volume construction request for the disk
+            let disk_volume = osagactx
+                .datastore()
+                .volume_get(disk.volume_id)
+                .await
+                .map_err(ActionError::action_failed)?;
+
+            warn!(
+                log,
+                "disk volume construction request {}",
+                disk_volume.data()
+            );
+
+            let disk_volume_construction_request: VolumeConstructionRequest =
+                serde_json::from_str(&disk_volume.data()).map_err(|e| {
+                    ActionError::action_failed(Error::internal_error(&format!(
+                        "failed to deserialize disk {} volume data: {}",
+                        disk.id(),
+                        e,
+                    )))
+                })?;
+
+            // Send the snapshot request to a random sled agent
+            let sled_agent_client = osagactx
+                .random_sled_client()
```

Comment:
I know that we don't have the most polished version of "sleds vs racks vs AZs", but at the moment, this looks like it could sent a request to *any sled in the fleet*. Is that what you want?

Reply:
The idea behind this was that the request will hit this Nexus, and *could* be handled by any sled agent because the disk is not currently attached to a instance. It doesn't matter *which* sled agent handles this request, and there's no notion of "the sled agent where this Nexus zone is" or "the sled agent closest to the crucible zone hosting the disk" anything like that.

In the future, I imagine this being restricted to "the AZ where the disk is" or something like that, yeah. Ultimately my understanding is that this request will be sent to the new service that @jclulow is writing, where the locality of that service will matter.